pkg/controller/certificaterequests/selfsigned/selfsigned.go

```
@@ -0,0 +1,173 @@
+	cmerrors "github.com/jetstack/cert-manager/pkg/util/errors"
+	"github.com/jetstack/cert-manager/pkg/util/kube"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	CRControllerName = "certificaterequests-issuer-selfsigned"
+)
+
+var (
+	errorNoAnnotation = fmt.Errorf("self signed issuer requires %q annotation to be set to the name of the Secret containing the private key",
+		v1alpha1.CRPrivateKeyAnnotationKey)
+)
+
+type SelfSigned struct {
+	// used to record Events about resources to the API
+	recorder record.EventRecorder
+
+	issuerOptions controllerpkg.IssuerOptions
+	secretsLister corelisters.SecretLister
+}
+
+func init() {
+	// create certificate request controller for selfsigned issuer
+	controllerpkg.Register(CRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		selfsigned := NewSelfSigned(ctx)
+		controller := certificaterequests.New(apiutil.IssuerSelfSigned, selfsigned)
+
+		c, err := controllerpkg.New(ctx, CRControllerName, controller)
+		if err != nil {
+			return nil, err
+		}
+
+		return c.Run, nil
+	})
+}
+
+func NewSelfSigned(ctx *controllerpkg.Context) *SelfSigned {
+	return &SelfSigned{
+		recorder:      ctx.Recorder,
+		issuerOptions: ctx.IssuerOptions,
+		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+	}
+}
+
+func (s *SelfSigned) Sign(ctx context.Context, cr *v1alpha1.CertificateRequest, issuerObj v1alpha1.GenericIssuer) (*issuer.IssueResponse, error) {
+	log := logf.FromContext(ctx, "sign")
+	reporter := crutil.NewReporter(cr, s.recorder)
+
+	resourceNamespace := s.issuerOptions.ResourceNamespace(issuerObj)
+
+	secretName, ok := cr.ObjectMeta.Annotations[v1alpha1.CRPrivateKeyAnnotationKey]
+	if !ok || secretName == "" {
+		message := fmt.Sprintf("Annotation %q missing or reference empty",
+			v1alpha1.CRPrivateKeyAnnotationKey)
+		err := errors.New("secret name missing")
+
+		reporter.Pending(err, "MissingAnnotation", message)
```

Undecided: should we hard fail if the annotation is missing? Do we want to allow changes here after the request has been created? I am leaning towards hard failing here 🙄

---

Yeah that's my bad - I agree, should hard fail